oauth2_client/client.py

```diff
@@ -1,37 +1,95 @@
 """
-Service to service call with OAuth2 authentication.
+HTTP(S) Session client, capable of running an OAuth2 authentication flow to
+obtain a token from an authorization provider.
+
+Enables authenticated communication between HTTP(s)-capable clients and servers.
+Identification is by means of OAuth token. Tokens are acquired by presentation
+of credentials, and may be refreshed by repeating the authorization flow.
+
+Supported OAuth2 grant types:
+- Client Credentials (e.g. service-to-service communication)
+- JWT Bearer token (e.g. Salesforce)
 """
-import time
+import logging
 
+import pybreaker
+from oauthlib.oauth2 import TokenExpiredError
 from requests_oauthlib import OAuth2Session
-from oauthlib.oauth2 import BackendApplicationClient
 
 from oauth2_client.compat import urljoin
-from .models import AccessToken, Application
+from oauth2_client.fetcher import fetch_token
+from oauth2_client.models import AccessToken, Application
+
+log = logging.getLogger(__name__)
 
+# Protect integration point with authorization provider
+token_breaker = pybreaker.CircuitBreaker(fail_max=3, reset_timeout=10)
 
-# Used to prevent the token expiration when the request is processed on the server if the token is expiring in seconds
-# or less at the client request time.
-TIMEOUT_SECONDS = 60
+# Protect integration point with communication receiver (resource owner)
+request_breaker = pybreaker.CircuitBreaker(fail_max=2, reset_timeout=10)
 
 
 class OAuth2Client(OAuth2Session):
     """
-    OAuth2 client to make HTTP calls with OAuth2 token.
+    OAuth2 client to make authorized HTTP(S) requests with OAuth2 token.
 
-    Wrapper around OAuth2Session to cache saervice host on the class instatiation and use it to transform
+    Wrapper around OAuth2Session to cache service host on the class instantiation and use it to transform
     relative urls to absolute ones.
 
+    Expired OAuth2 token can be refreshed by repeating the authorization flow. Token refresh
+    functionality is protected with circuit breakers, to avoid spamming authorization provider
+    and resource owner in case of system failure. Token expiry detection uses best-effort approach,
+    as there is no standardized way of informing the client of token expiry. The mechanism tries
```

Comment:
Because token expiration data is not always set?

Reply:
resolved on a call